crawler/priority_samples_process.py

```diff
@@ -0,0 +1,227 @@
+
+
+def update_priority_samples(db: Database, config: Config, add_to_dart: bool) -> None:
+    """
+    Update any remaining unprocessed priority samples in MLWH and DART
+    Arguments:
+        db {Database} -- mongo db instance
+        config {Config} -- config for mysql and dart connections
+    """
+
+    def extract_mongo_id(sample: SampleDoc) -> SampleDocValue:
+        return sample[FIELD_MONGODB_ID]
+
+    logger.info("Starting Step 2")
+
+    samples = query_any_unprocessed_samples(db)
+
+    # Create all samples in MLWH with samples containing both sample and priority sample info
+    mlwh_success = update_priority_samples_into_mlwh(samples, config)
+
+    # Add to the DART database if MLWH update was successful
+    if mlwh_success:
+        logger.info("MLWH insert successful")
+        if add_to_dart:
+            logger.info("Adding to DART")
+            dart_success = insert_plates_and_wells_into_dart(samples, config)
+        if not (add_to_dart) or dart_success:
+            logger.info("Updating Mongodb priority samples to processed")
+            # use stored identifiers to update priority_samples table to processed true
+            sample_ids = list(map(extract_mongo_id, samples))
+            update_unprocessed_priority_samples_to_processed(db, sample_ids)
+
+
+def query_any_unprocessed_samples(db: Database) -> List[SamplePriorityDoc]:
+    priority_samples_collection = get_mongo_collection(db, COLLECTION_PRIORITY_SAMPLES)
+
+    # Query:
+    # get all unprocessed priority samples, we want to update also samples that have changed their importance
+    # join on the samples collection using sample_id in priority_samples collection to _id in the samples collection
+    # flatten object so sample fields are at the same level as the priority samples fields, removing
+    # the nested sample object return a list of the samples
+    # e.g db.priority_samples.aggregate([{"$match":{"$and": [
+    # {"processed": false} ]}}, {"$lookup": {"from": "samples", "let": {"sample_id": "$_id"},"pipeline":
+    # [{"$match": {"$expr": {"$and":[{"$eq": ["$sample_id", "$$sample_id"]}]}}}], "as": "from_samples"}}])
+    IMPORTANT_UNPROCESSED_SAMPLES_MONGO_QUERY: Final[List[object]] = [
+        {
+            "$match": {
+                "$and": [
+                    {FIELD_PROCESSED: False},
+                ]
+            }
+        },
+        {
+            "$lookup": {
+                "from": "samples",
+                "localField": FIELD_SAMPLE_ID,
+                "foreignField": FIELD_MONGODB_ID,
+                "as": "related_samples",
+            }
+        },
+        {"$match": {"related_samples": {"$ne": []}}},
+        {"$replaceRoot": {"newRoot": {"$mergeObjects": [{"$arrayElemAt": ["$related_samples", 0]}, "$$ROOT"]}}},
+        {"$project": {"related_samples": 0}},
+    ]
+
+    value = priority_samples_collection.aggregate(IMPORTANT_UNPROCESSED_SAMPLES_MONGO_QUERY)
+    return list(value)
```

Comment:
Can we prevent converting it to a list and using it as an iterator or is that a larger refactor? I think it will be more efficient as an iterator.

Reply:
Yeah, thought it might be quite a large refactor, thanks for the update.